### PR TITLE
Edit post tags

### PR DESCRIPTION
**Pull request by tomato-gits**
- anyone can edit tags on their own posts with the regular edit feature
- teachers can edit tags on other posts (student or teacher)
- when teachers edit tags on posts they did not author, they are limited by the post author's role (this is different than what I planned. it looked like editing posts from the student side was going to be complicated if a teacher tagged their post with a teacher-only tag)  maybe we can decide together if it's worth the effort to let teachers use any tag on student posts

PostList.vue shows the list of all posts, this is where the tag-badge needs to be updated after a post's tags have been updated

PostDisplay.vue has new pieces that track closely with the post edit feature already in place. can_edit_tag is a new computed prop for ability to edit just tags, open_tag_editor and tag_save_pending handle the form display, and edit_tag_options filters available tags based on author's role (see the note above, we may decide to do this differently)
handle_close_post_editor_ok was already handling clearing out the form data for the edit post form, this now also clears out the tag edit form

store.js has a new editTag action that calls the new route in api.php, this gets the ApiController function editTag (little bit of different auth logic than the existing editPost handler)

### CODE REVIEW

File: web/resources/js/components/PostDisplay.vue [hunk 1]
```diff
@@ -37,6 +42,11 @@ export default {
             return this.post.author_user_id === this.$store.getters.user.id
                 && !this.course_is_closed
         },
+        can_edit_tag() {
+            return this.user_is_teacher
+                && this.post.author_user_id !== this.$store.getters.user.id
+                && !this.course_is_closed
+        },
```
Comment (chrispittman) [inline web/resources/js/components/PostDisplay.vue:49]: This is "only teachers can edit tags, and only on other people's posts".  Is that the correct condition here?

File: web/resources/js/components/PostDisplay.vue [hunk 2]
```diff
@@ -58,7 +68,23 @@ export default {
...
+            return post_tags;
+        },
+        edit_tag_options() {
+            let author_is_teacher = this.post.author_user_role === 'teacher'
+            let post_tags = this.$store.getters.course_summary.post_tags
+                .filter(t => {return author_is_teacher || !t.teacher_only})
+                .map(t => {return { value: t.name, text: t.name }})
+            post_tags.unshift({ value: null, text: '(no tag)' })
+            return post_tags;
+        },
```
Comment (chrispittman) [inline web/resources/js/components/PostDisplay.vue:87]: Do you think it's worth extracting out a `get_tags_available_for_role()` function, and changing this to:
```
    post_tag_options() {
         return get_tags_available_for_role(this.$store.getters.user.role);
    },
    edit_tag_options() {
         return get_tags_available_for_role(this.post.author_user_role);
    }
```
...?

File: web/resources/js/components/PostDisplay.vue (hunk 1 above)
Comment (tomato-gits) [reply web/resources/js/components/PostDisplay.vue:49]: Technically the only rule is that "only teachers can edit tags", but as long as there are separate edit forms and menu links for editing a post vs just the tag, this extra condition makes the UI simpler.   
If this condition just said "only teachers can edit tags", then teachers would see 2 menu options (see screenshot) and the editing form would look different depending on which one they picked. We could do a little extra display logic on those buttons/forms; adding the condition about post.author just shortcuts that.
Screenshot 2024-04-04 at 2 28 58 PM

**Review by chrispittman — APPROVED** [review]
LGTM 👍

**Review by sunil-nar — APPROVED** [review]

**Pull request merged**